### PR TITLE
define separate semantics for `$`, and `[[`

### PR DESCRIPTION
**Pull request by kevinushey**
Related: https://github.com/rstudio/reticulate/issues/251

This PR seeks to define different semantics for `$` and `[[`. With this PR:

- `$` is used for attribute access, as Python's `.` is used;
- `[[` is used for item access, as Python's `[[` is used.

For backwards compatibility, with Python (builtin) dictionaries, `$` still returns items rather than attributes. We may want to consider changing this in the future.

### CODE REVIEW

**Comment by kevinushey** [comment]
For what it's worth, some of the surprise here might've arisen from how Pandas defines the `.` operator. For Pandas DataFrames, `.` will first look at attributes; if no such attribute exists, then it will attempt to look for a column. This implies that code of this form works:

```
>>> import pandas as pd
>>> pdf = pd.DataFrame({'apple': [1]})
>>> pdf.apple
0    1
Name: apple, dtype: int64
```

However, if an attribute (method) of the same name exists, then that is chosen instead:

```
>>> import pandas as pd
>>> pdf = pd.DataFrame({'pop': [1]})
>>> pdf.pop
<bound method DataFrame.pop of    pop
0    1>
```

**Comment by kevinushey** [comment]
To help preserve backwards compatibility, I've decided on the following:

- `$` accesses attributes first, items second;
- `[` accesses items first, attributes second.

This should ensure the least amount of breakage in existing code, but it does imply a change in this case for example:

```
> d <- dict(items = 42)
> d$items
<built-in method items of dict>
> d['items']
42.0
```

Previously, `d$items` would have accessed the `items` item; now it accesses the attribute (bound method) called `items`.

File: R/python.R [hunk 1]
```diff
@@ -231,41 +250,55 @@ py_has_convert <- function(x) {
...
-  else if (inherits(x, "python.builtin.dict"))
-    attrib <- py_dict_get_item(x, name)
-  else
-    attrib <- py_get_attr(x, name)
+  # get the attrib and convert as needed
+  if (prefer_attr) {
 
-  # convert
-  if (convert || py_is_callable(attrib)) {
+    if (py_has_attr(x, name)) {
```
Comment (kevinushey) [inline R/python.R]: These checks should only occur if the user is passing in a scalar string (length one character vector).

**Comment by jjallaire** [comment]
@kevinushey Is this good to merge?

**Comment by kevinushey** [comment]
I want to double-check this doesn't regress anything in keras or tensorflow so I'll report back after testing a bit there.

**Comment by jjallaire** [comment]
Okay, you can go ahead and merge whenever you think it's ready.

**Comment by kevinushey** [comment]
Tests pass with both `keras` and `tensorflow` with this branch, so I'm going to merge.

**Pull request merged**